### ISSUE TITLE
Syntax for flat Modelica output in the case of ragged arrays

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#8198]
Consider the attached [TestArray.zip](https://github.com/OpenModelica/OpenModelica/files/7584707/TestArray.zip) package.

When flattened, `TestArray.ConnectEquations.SC` has a 3D rectangular array defined as:
```Modelica
  Real[3,3] c.x;
```
Consider now `TestArray.ConnectEquations.SD`. The flat model contains a ragged array `c.x`. The NF already handles it (we had to implement this feature for the Buildings library), but it cannot output it in textual form when `-d=nonfScalarize` is activated.

I would propose to extend the standard Modelica syntax for rectangular arrays, by substituting the second index, which in the rectangular array has the same range for each value of the first index, with an array with the maximum values of the individual ranges for each value of the first index. Specifically, in the case of `TestArray.ConnectEquations.SD` one would get:
```Modelica
  Real[3, {3, 4, 5}] c.x;
```
Literal ragged array constants can be obviously defined as arrays of arrays, with different numbers of elements, e.g.:
```Modelica
  Real[3, {3, 4, 5}] c.p = 
    { {2, 2, 3}
      {2, 3, 2, 4}
      {1, 4, 3, 2, 5}
    };
```
The approach can be readily generalized to multi-dimensional ragged arrays: for each element of the i-th dimension of the array, we put in the i-1-th dimension one array with the corresponding dimensions, e.g.:
```Modelica
  Real[3, {3, 4, 10}, {{2,2,3}, {2,3,2,4}, {2,2,2,2,2,2,2,2,5,1}}] c.m.x;
```
If some dimensions after the first are equal to each other, we will just put a scalar index instead of an array, e.g.:
```Modelica
  Real[2, {3, 2), 2] c.m.x;
```
would be a 3D ragged array where the 3rd dimension has the same range for all the elements, while
```Modelica
  Real[2, {3, 2), {2, 3}] c.m.x;
```
would contain the following scalar elements:
```Modelica
  c.m.x[1, 1, 1]
  c.m.x[1, 1, 2]
  c.m.x[1, 2, 1]
  c.m.x[1, 2, 2]
  c.m.x[1, 3, 1]
  c.m.x[1, 3, 2]
  c.m.x[2, 1, 1]
  c.m.x[2, 1, 2]
  c.m.x[2, 1, 3]
  c.m.x[2, 2, 1]
  c.m.x[2, 2, 2]
  c.m.x[2, 2, 3]
```
or, as another example:
```Modelica
  Real[2, {3, 2), {{2, 4, 2}, 3}] c.m.x;
```
would expand into
```
  c.m.x[1, 1, 1]
  c.m.x[1, 1, 2]
  c.m.x[1, 2, 1]
  c.m.x[1, 2, 2]
  c.m.x[1, 2, 3]
  c.m.x[1, 2, 4]
  c.m.x[1, 3, 1]
  c.m.x[1, 3, 2]
  c.m.x[2, 1, 1]
  c.m.x[2, 1, 2]
  c.m.x[2, 1, 3]
  c.m.x[2, 2, 1]
  c.m.x[2, 2, 2]
  c.m.x[2, 2, 3]
```
@perost, if this makes sense we could implement it, and then propose it to the Flat Modelica working group as an extension to support non-expanded array backends.

I cc: @ennioquaglia for his information.

**Comment by perost** [comment]
If a remove the for-loop in `TestArray.ConnectEquations.CC` the model flatten to:
```Modelica
class TestArray.ConnectEquations.SD "A model using arrays of models with arrays of basic types (ragged array)"
  final parameter Integer N = 3;
  parameter Real[3] p = {1.0, 1.5, 2.0};
  Real[3] c.c.f;
  Real[3] c.c.e;
  Real[3, {3, 4, 5}] c.x;
  parameter Real[3] c.p = p;
  final parameter Integer[3] c.N = {3, 4, 5};
  parameter Real s.p = 3.0;
  Real s.c.e;
  Real s.c.f;
equation
  c.c.f = 0.0;
  c[3].c.e = c[2].c.e;
  c[3].c.e = c[1].c.e;
  c[3].c.e = s.c.e;
  c[3].c.f + c[2].c.f + c[1].c.f + s.c.f = 0.0;
  c.x[1] = c.c.e;
  c.x[c.N] = c.c.f;
  s.c.f = s.p;
end TestArray.ConnectEquations.SD;
```
So that's how we already handle it. The issue with the for-loop is that it turns into:
```Modelica
  for $i1 in 1:3 loop
    for i in 2:{3, 4, 5}[$i1] loop
      c[$i1].x[i] = c[$i1].x[i - 1] + c[$i1].p;
    end for;
  end for;
```
And then we try to simplify the type of the range, which fails because neither the constant evaluation nor the simplification module expects dimensions to be arrays.

**Comment by casella** [comment]
> The issue with the for-loop is that it turns into:
> 
> ```modelica
>   for $i1 in 1:3 loop
>     for i in 2:{3, 4, 5}[$i1] loop
>       c[$i1].x[i] = c[$i1].x[i - 1] + c[$i1].p;
>     end for;
>   end for;
> ```

LGTM. @ennioquaglia, would this be ok for you?
 
> And then we try to simplify the type of the range, which fails because neither the constant evaluation nor the simplification module expects dimensions to be arrays.

Could we just have a flag to stop short of that, and output what you showed above?

**Comment by ennioquaglia** [comment]
Yes, seems good to me.

**Comment by perost** [comment]
#8203 fixes the simplification issue by retyping ranges when we evaluate any constant/structural parameter in them. I think this is a good idea in general, because a range like `start:stop` at first gets a type like `Integer[max(stop - start + 1, 0)]`. If we evaluate anything in the range it's less work to just recalculate the type instead of evaluating the type, and that also avoids the issue with `nonfScalarize` since the type otherwise gets messed up.

Now the model fails because `x` in `c[$i1].x[i]` inside the loop has the type `[{3, 4, 5}[<c, 1>]]`, where `<c, 1>` is a subscript that should normally be replaced with the first subscript of the `c`-element we're flattening. But since this stage is skipped when using `nonfScalarize` it's never replaced and fails when trying to convert it to DAE format. I think the correct solution is probably to replace the subscript with `$i1` in that case, but `nonfScalarize` is doing some weird string matching to figure out what to replace rather than the node matching that the normal flattening does. If you need it I can probably find some way to fix that issue though.

I also opened #8204 which adds a new flag `--dumpFlatModel` that replaces the old debug flag `-d=nfDumpFlat` (which no one but me probably knew about). The difference is that `--dumpFlatModel` can be used to dump the flat model at chosen stages of the flattening, while `-d=nfDumpFlat` always dumped it at the end. The old flag would probably have been good enough for you, but I wanted the new one anyway since it's useful for debugging.

The flag can be use like `--dumpFlatModel=all` to dump after every stage, or `--dumpFlatModel=stage1,stage2,etc` to dump after only chosen stages. The stages are currently (in order of execution):
```
 * flatten                    After flattening but before connection handling.
 * connections                After connection handling.
 * eval                       After evaluating constants.
 * simplify                   After model simplification.
 * scalarize                  After scalarizing arrays.
 * all                        Dump after every stage.
```
Note that this will only print the model and not the used functions though. The reason is simply that we haven't actually implemented a print function for functions in the NF, but if you need it I can add it (we do have a function for Flat Modelica that I can copy and modify, so it's probably not much work). We currently collect the used functions after the simplification step though, so you'd only get them for the `simplify` and `scalarize` stages.

**Comment by casella** [comment]
> Now the model fails because `x` in `c[$i1].x[i]` inside the loop has the type `[{3, 4, 5}[<c, 1>]]`, where `<c, 1>` is a subscript that should normally be replaced with the first subscript of the `c`-element we're flattening. But since this stage is skipped when using `nonfScalarize` it's never replaced and fails when trying to convert it to DAE format. I think the correct solution is probably to replace the subscript with `$i1` in that case, but `nonfScalarize` is doing some weird string matching to figure out what to replace rather than the node matching that the normal flattening does. If you need it I can probably find some way to fix that issue though.

As far as I am concerned, the goal is to get
```Modelica
  for $i1 in 1:3 loop
    for i in 2:{3, 4, 5}[$i1] loop
      c[$i1].x[i] = c[$i1].x[i - 1] + c[$i1].p;
    end for;
  end for;
```
in the flat output. I guess I can't help you much suggesting how to get there 😄. I only hope it is not too hard, otherwise we could examine other options.

> I also opened #8204 which adds a new flag `--dumpFlatModel` that replaces the old debug flag `-d=nfDumpFlat` (which no one but me probably knew about). The difference is that `--dumpFlatModel` can be used to dump the flat model at chosen stages of the flattening, while `-d=nfDumpFlat` always dumped it at the end. The old flag would probably have been good enough for you, but I wanted the new one anyway since it's useful for debugging.

LGTM. I guess we probably want to get at least to the connection phase. The other steps we could do ourselves. Do I understand correctly that instantiateModel() coupled to --nonfScalarize is the same as -dumpFlatModel = simplify?

@ennioquaglia please take note, information in the intermediate stages may be interesting or useful to you, at least for discussion.

> Note that this will only print the model and not the used functions though. The reason is simply that we haven't actually implemented a print function for functions in the NF, but if you need it I can add it (we do have a function for Flat Modelica that I can copy and modify, so it's probably not much work). We currently collect the used functions after the simplification step though, so you'd only get them for the `simplify` and `scalarize` stages.

I guess it would be nice to also have them before, for consistency. Only if it is not too much work.

**Comment by perost** [comment]
> LGTM. I guess we probably want to get at least to the connection phase. The other steps we could do ourselves. Do I understand correctly that instantiateModel() coupled to --nonfScalarize is the same as -dumpFlatModel = simplify?

We handle variables with 0-size dimensions after the simplify stage regardless of if scalarization is on or not, removing the variables and replacing references to them with empty arrays expressions. So the model might change a bit after `simplify` even if scalarization is turned off, but only if it contains variables with 0-size dimensions.

Also, in case I wasn't clear the example model can be dumped at every stage now that I've fixed the simplification issue. It only fails in the conversion to DAE, so the normal flat model can't be printed since it's based on the DAE.

**Comment by casella** [comment]
> Also, in case I wasn't clear the example model can be dumped at every stage now that I've fixed the simplification issue. It only fails in the conversion to DAE, so the normal flat model can't be printed since it's based on the DAE.

Ah, OK, I missed that. 

So, I shouldn't rely on what I get by clicking on the Instantiate Model button in OMEdit. I'm not sure I understand exactly what are the steps to get the right output for our purpose. After #8203 I guess it's going to be the following ones:
```
setCommandLineOptions("-d=arrayConnect,combineSubscripts,printRecordTypes --showStructuralAnnotations --dumpFlatModel=simplify --flatModelica");
```
do I get this right?

One last thing. Suppose we are flattening a model with an array that has 10 million elements. Not only we want a flat model which is not scalarized, but we also want to avoid the scalarization step entirely, otherwise it could take forever to run it, and there is no point doing that. If we use `--dumpFlatModel=simplify`, is the scalarization step actually skipped, or simply not reported? Should we combine that flag with `-d=nonfScalarize` to skip it?

**Comment by perost** [comment]
> > Also, in case I wasn't clear the example model can be dumped at every stage now that I've fixed the simplification issue. It only fails in the conversion to DAE, so the normal flat model can't be printed since it's based on the DAE.
> 
> Ah, OK, I missed that.
> 
> So, I shouldn't rely on what I get by clicking on the Instantiate Model button in OMEdit. I'm not sure I understand exactly what are the steps to get the right output for our purpose. After #8203 I guess it's going to be the following ones:
> 
> ```
> setCommandLineOptions("-d=arrayConnect,combineSubscripts,printRecordTypes --showStructuralAnnotations --dumpFlatModel=simplify --flatModelica");
> ```
> 
> do I get this right?

`--flatModelica` (or `-f` for short) won't work either if the DAE conversion fails. It's based on the NF structure though, so it's possible to print it even in that case, but the way the code is structured at the moment means that the DAE conversion is done before printing the flat model. I guess we could add a flag for only running the frontend, which could skip the DAE conversion if `-f` is used since it's only need by the backend in that case.

Also, `--dumpFlatModel` and `--f` are not using the same format, `--dumpFlatModel` uses the NF format which mimics the normal flat output while `--f` uses the new experimental Flat Modelica format. I can change `--dumpFlatModel` to use Flat Modelica format if `-f` is used though, I didn't think of that but it's trivial to do.

> One last thing. Suppose we are flattening a model with an array that has 10 million elements. Not only we want a flat model which is not scalarized, but we also want to avoid the scalarization step entirely, otherwise it could take forever to run it, and there is no point doing that. If we use `--dumpFlatModel=simplify`, is the scalarization step actually skipped, or simply not reported? Should we combine that flag with `-d=nonfScalarize` to skip it?

`--dumpFlatModel` does not affect the flattening at all, it just dumps the flat model at chosen stages. It's kind of the equivalent of `-d=optdaedump` but for the frontend, except that you can choose which stages you want to dump.

**Comment by casella** [comment]
> `--flatModelica` (or `-f` for short) won't work either if the DAE conversion fails. It's based on the NF structure though, so it's possible to print it even in that case, but the way the code is structured at the moment means that the DAE conversion is done before printing the flat model. I guess we could add a flag for only running the frontend, which could skip the DAE conversion if `-f` is used since it's only need by the backend in that case.

Sounds good for us, we don't need the backend.

> Also, `--dumpFlatModel` and `--f` are not using the same format, `--dumpFlatModel` uses the NF format which mimics the normal flat output while `--f` uses the new experimental Flat Modelica format. I can change `--dumpFlatModel` to use Flat Modelica format if `-f` is used though, I didn't think of that but it's trivial to do.

Please do. The Flat Modelica format is going to be standardized and become part of eFMI, and I understand that part (the quoting) is there to stay, so even though it is "experimental" it's going to be more stable in the long term. Hence, I think it's best for us to use that from the start.

> `--dumpFlatModel` does not affect the flattening at all, it just dumps the flat model at chosen stages. It's kind of the equivalent of `-d=optdaedump` but for the frontend, except that you can choose which stages you want to dump.

Yeah, that's my point. How can be skip the flattening to scalar phases, which is useless for us and could take a lot of time for nothing?

**Comment by perost** [comment]
#8214 fixes the DAE conversion issue, so you can now get the normal flat model output without using `--dumpFlatModel`. But in case you want to get the flat model at some intermediate stage #8215 also improves `--dumpFlatModel` so that it uses the Flat Modelica format when `-f` is used.

The output when using `-d=-nfScalarize,arrayConnect,combineSubscripts,printRecordTypes -f` for the example model is now:
```Modelica
class 'TestArray.ConnectEquations.SD'
function 'subscriptInteger[3]_Integer'
  input Integer[3] exp;
  input Integer s1;
  output Integer result = exp[s1];
end 'subscriptInteger[3]_Integer';

  public parameter Integer 'N' = 3;
  public parameter Real[3] 'p' = {1.0, 1.5, 2.0};
  public Real[3] 'c.c.f';
  public Real[3] 'c.c.e';
  public Real[3, {3, 4, 5}] 'c.x';
  public parameter Real[3] 'c.p' = 'p'[:];
  public parameter Integer[3] 'c.N' = {3, 4, 5};
  public parameter Real 's.p' = 3.0;
  public Real 's.c.e';
  public Real 's.c.f';
public
equation
  'c.x'[:,1] = 'c.c.e'[:];
  'c.x'[:,'c.N'[:]] = 'c.c.f'[:];

  for '$i1' in 1:3 loop
    for 'i' in 2:'subscriptInteger[3]_Integer'({3, 4, 5},'$i1') loop
      'c.x'['$i1','i'] = 'c.x'['$i1','i' - 1] + 'c.p'['$i1'];
    end for;
  end for;

  's.c.f' = 's.p';
  'c.c.e'[2] = 'c.c.e'[1];
  's.c.e' = 'c.c.e'[1];
  'c.c.e'[3] = 'c.c.e'[1];
  'c.c.f'[3] + 's.c.f' + 'c.c.f'[2] + 'c.c.f'[1] = 0.0;
end 'TestArray.ConnectEquations.SD';
```
The weird function `'subscriptInteger[3]_Integer'` is a subscripting function. This is generated because we wanted the Flat Modelica to be legal Modelica, and Modelica does not allow subscripting generic expressions (which is common in flat models). So I solved it by generating subscripting functions, which basically converts an expression to a cref that can be subscripted.

It seems Flat Modelica won't be legal Modelica though, and will allow subscripting generic expression. So if we implement the actual Flat Modelica format those subscripting functions won't be needed. For now they are needed though, but if they cause issues for you we'll have to discuss what to do about them.

**Comment by casella** [comment]
> #8214 fixes the DAE conversion issue, so you can now get the normal flat model output without using `--dumpFlatModel`. But in case you want to get the flat model at some intermediate stage #8215 also improves `--dumpFlatModel` so that it uses the Flat Modelica format when `-f` is used.

Excellent!

> The weird function `'subscriptInteger[3]_Integer'` is a subscripting function. This is generated because we wanted the Flat Modelica to be legal Modelica, and Modelica does not allow subscripting generic expressions (which is common in flat models). So I solved it by generating subscripting functions, which basically converts an expression to a cref that can be subscripted.
> 
> It seems Flat Modelica won't be legal Modelica though, and will allow subscripting generic expression. So if we implement the actual Flat Modelica format those subscripting functions won't be needed. For now they are needed though, but if they cause issues for you we'll have to discuss what to do about them.

This is largely a matter of philosophical discussion 😄 

Seriously, the "default" flat output is not legal Modelica anyway, because as far as I understand, you can't even have declarations with dots in a variable name, like
```
  Real a.b.x;
```
so you can't pass that stuff to a Modelica compiler anyway. That said, we are not really building a full Modelica compiler, though, currently we are planning to rely on the OMC NF to provide us some kind of flat output. Hence, in principle, whatever the NF passes us is good, as long as we agree on it. My preference would be to stick to the Flat Modelica standard, because that's, well, a standard, while any other dialect we could imagine (including legal Modelica ones) aren't, and we should probably focus on implementing one standard, rather than 10 different dialects for 10 different users.

Summing up, for me it's fine that the -f output has a subscripted expression, if that is legal according to the draft of the Flat Modelica standard. We don't need any weird function to make it legal Modelica, because we don't need a legal Modelica output, just a standard flat Modelica one.

@ennioquaglia any thoughts on that, since you're going to implement stuff based on that?

**Comment by perost** [comment]
> Summing up, for me it's fine that the -f output has a subscripted expression, if that is legal according to the draft of the Flat Modelica standard. We don't need any weird function to make it legal Modelica, because we don't need a legal Modelica output, just a standard flat Modelica one.

The issue is that we implemented the current Flat Modelica prototype for some project that needed it to be legal Modelica. Obviously it doesn't work in this case since e.g. `{3, 4, 5}` is not a valid dimension, but for normal cases it does produce legal Modelica (well, most of the time, when things work properly...).

I'm not sure if this is still needed though, or if we should target actual Flat Modelica more and just output subscripted expressions.

@sjoelund: What do you think?

**Comment by sjoelund** [comment]
I think if we could allow general subscripts on an experimental flag, that would be enough

**Comment by perost** [comment]
> I think if we could allow general subscripts on an experimental flag, that would be enough

The NF pretty much already supports it, but we'd have to add support for it in the parser and Absyn/SCode which is a bit of work. The easiest would be to just add another flag to disable the current behaviour, but we're already drowning in flags as it is.

**Comment by casella** [comment]
I'm not sure what project that was, and what were the requirements for it, maybe you can give me more information. My point is that as of Nov 2021, "Flat Modelica" is no longer a vague concept, but is going towards being a standard for Flat Modelica output, see [MCP 0031](https://github.com/modelica/ModelicaSpecification/tree/MCP/0031/RationaleMCP/0031). This will take some time, but the direction is set, and what has already been decided won't likely get undone in the future.

As I understand it, the idea of Flat Modelica is to allow using tools such as the new front end of OMC to take the burden of flattening off other tools (including our LLVM compiler at Polimi) that will instead take care of lower-level code generation issues. As OMC now pretty much covers 100% of Modelica 3.5 (except state machines), it could actually get very widely used for this purpose, which would be good for the OSMC. Of course that only works if we stick to the standard, as we always do.

As such, for the time being, I don't see any need for OMC to be able to _read_ Flat Modelica code, as it can now read Modelica code quite well. So, there is no need for @perost to do add support for that stuff in the parser, absyn and scode.

To the contrary, I do see a need for OMC to be able to _output_ Flat Modelica code, so that other tools could use it to do stuff, within the context of eFMI, or possibly outside it, as in our case at Polimi. In order to do that, we shouldn't need to turn on half a dozen flags to actually get the output according to the standard, `-f` should do. Extra flags would only be needed for non-scalarized output, which is still not defined in MCP 0031 (as far as I know) and is still highly experimental.

So, my proposal is to activate expression subscripting _output_ with `-f` and to have a flag with the weird function call for backwards-compatibility of the previous project.

**Comment by sjoelund** [comment]
How would we verify that the flat code is correct if we cannot run it? That's the main reason for supporting it

**Comment by casella** [comment]
> How would we verify that the flat code is correct if we cannot run it? That's the main reason for supporting it

The main reason to support Flat Modelica is to provide flat Modelica code to downstream tools. Using it makes little sense per se, since we already have an excellent regular Modelica front end.

I agree that verifying the correctness of the flat Modelica code by running it and comparing the results with what we'd get directly is one important use case, but certainly not the main one. In any case, it's not needed _urgently_. 

@perost will eventually need to write the parser for Flat Modelica, since it is our mission to support all the MA standards fully, but that's not on the critical path right now. For the simple examples that we need for our LLVM backend, we can check the correctness of the code by inspection on simple enough cases. On more complex ones, we'll run simulation and compare them with the native OMC one.

**Comment by sjoelund** [comment]
> I agree that verifying the correctness of the flat Modelica code by running it and comparing the results with what we'd get directly is one important use case, but certainly not the main one. In any case, it's not needed urgently.

Untested code dies. It's very easy to introduce breaking changes if it's not tested. If you now say you want to build other tools on top of this, it for sure needs to be tested.

**Comment by casella** [comment]
I couldn't agree more. For the time being, the test can consist in flattening the examples of the attached package (we can and will add more for our own testing purposes), check the correctness of the output manually (we already did that, and now it seems to be fine, except for arrayConnect, that only works in some cases, see #8014), and then just check that the output remains the same and does not change.

Would this work for you?

**Comment by perost** [comment]
I remembered that we have an old flag `--modelicaOutput` (`-m`) whose description is literally `"Enables valid modelica output for flat modelica."`. So I made the output of subscripting functions optional in #8224. By default `-f` now outputs subscripted expression, and if you instead want subscripting functions you can use `-f -m`.

**Comment by sjoelund** [comment]
Yeah, this seems like a reasonable solution with sufficiently low cost of implementation

**Comment by casella** [comment]
LGTM. 

@perost, does the merging in of #8224 close this issue? I can't tell until tomorrow's nightly becomes available, but I guess so.

**Comment by perost** [comment]
> LGTM.
> 
> @perost, does the merging in of #8224 close this issue? I can't tell until tomorrow's nightly becomes available, but I guess so.

Maybe, that depends on if the current output works for you or if there's still something that needs to be changed.

**Comment by casella** [comment]
I just flattened `TestArray.ConnectEquations.SD` with `-f -d=nonfScalarize,arrayConnect,combineSubscripts,printRecordTypes --newBackend --showStructuralAnnotations` using the latest nightly build and I got
```
class 'TestArray.ConnectEquations.SD'
  public parameter Integer 'N' = 3;
  public parameter Real[3] 'p' = {1.0, 1.5, 2.0};
  public Real[3] 'c.c.f';
  public Real[3] 'c.c.e';
  public Real[3, {3, 4, 5}] 'c.x';
  public parameter Real[3] 'c.p' = 'p'[:];
  public parameter Integer[3] 'c.N' = {3, 4, 5};
  public parameter Real 's.p' = 3.0;
  public Real 's.c.e';
  public Real 's.c.f';
public
equation
  'c.x'[:,1] = 'c.c.e'[:];
  'c.x'[:,'c.N'[:]] = 'c.c.f'[:];

  for '$i1' in 1:3 loop
    for 'i' in 2:{3, 4, 5}['$i1'] loop
      'c.x'['$i1','i'] = 'c.x'['$i1','i' - 1] + 'c.p'['$i1'];
    end for;
  end for;

  's.c.f' = 's.p';
  'c.c.e'[2] = 'c.c.e'[1];
  's.c.e' = 'c.c.e'[1];
  'c.c.e'[3] = 'c.c.e'[1];
  'c.c.f'[3] + 's.c.f' + 'c.c.f'[2] + 'c.c.f'[1] = 0.0;
end 'TestArray.ConnectEquations.SD';
```
which is exactly what we expected to get. @ennioquaglia please confirm. Thanks @perost!
